ql/src/java/org/apache/hadoop/hive/ql/parse/BaseSemanticAnalyzer.java

```diff
@@ -2092,4 +2110,79 @@ protected String getFullTableNameForSQL(ASTNode n) throws SemanticException {
     }
   }
 
+  public QB getQB(){
+    return null;
+  }
+
+  /**
+   * Resolves the query type from the SemanticAnalyzer and the query AST, which might look magic.
+   * Magic is needed because a query can fail due to semantic exceptions, and at that time, from syntax point of view,
+   * it's already clear what's the query type, even if the SemanticAnalyzer bailed out, typical problems:
+   * 1. general semantic exception
+   * 2. transaction manager validation (throwin exception)
+   * What a user expects here is something like "QUERY", "DDL", "DML", so this magic will do its best to tell.
+   * any kind of "analyze": STATS
+   * DML operations (INSERT, UPDATE, DELETE, MERGE): DML
+   * MAPRED: QUERY, DML (depending on QueryProperties achieved in compile time)
+   * FETCH: QUERY, as a simple fetch task is a QUERY
+   * empty string if we can't determine the type of the query,
+   * e.g. when ParseException happens, we won't do further magic,
+   * even if it's obvious by reading the sql statement that user wanted to run e.g. a select query
+   * @param tree the root ASTNode of the query
+   */
+  public void setQueryType(ASTNode tree) {
+    List<Task<? extends Serializable>> rootTasks = getAllRootTasks();
+    if (queryProperties == null) {
+      return;
+    }
+    queryProperties.setQueryType(queryProperties.isAnalyze() ? QueryType.STATS :
+        (queryProperties.isDML() ?
+            QueryType.DML : null));
+    // common confusion whether CTAS is DML or DDL, let's pick DDL here
+    if (queryProperties.isCTAS()) {
+      queryProperties.setQueryType(QueryType.DDL);
+    }
+    if (queryProperties.getQueryType() != null) {
+      return; //already figured out
+    }
+    queryProperties.setQueryType(rootTasks.stream().findFirst().map(t -> {
+      String type = t.getType().toString();
+      // a MAPRED stage could mean an INSERT query also
+      if (queryProperties != null && type.equalsIgnoreCase("MAPRED")) {
+        return queryProperties.isDML() ? QueryType.DML : QueryType.QUERY;
+      }
+      return null;
+    }).orElseGet(() -> {
+      // in case of a semantic exception (e.g. a table not found or something else)
+      // the root AST Node can still imply if this is a query, try to fall back to that
+      // instead of ""
+      if (tree.getText().equalsIgnoreCase("TOK_QUERY")) {
```

Comment:
Generally it is better to compare the constant with the nullable value like 
```
"TOK_QUERY".equalsIgnoreCase(tree.getText())
```
It is null safe.

Another option is
```
StringUtils.equalsIgnoreCase(tree.getText(), "TOK_QUERY")
```